AIP 39: Documentation (#17552)
mistercrunch/Airflow: docs/apache-airflow/dag-run.rst

Your DAG will be instantiated for each schedule along with a corresponding
DAG Run entry in the database backend.

Data Interval
-------------

Each DAG run in Airflow has an assigned "data interval" that represents the time
range it operates in. For a DAG scheduled with ``@daily``, for example, each of

> its data interval would start at midnight of each day, and end at midnight of

Reviewer: ```suggestion
its data interval would start at midnight of each day and end at midnight of
```

the next day.

> A DAG run happens *after* its associated data interval has ended, to ensure the

Reviewer 2: ```suggestion
A DAG run created *after* its associated data interval has ended, to ensure the
```

Reviewer: ```suggestion
A DAG run is created *after* its associated data interval has ended, to ensure the
```

"happens" is fine, otherwise it should be "is created". The latter might be a better choice, but really either work.

Author: I was trying to avoid the passive “is created” because the data interval’s end does not technically trigger the run’s creation, but merely triggers a request to create a run. I’m not sure whether this distinction is meaningful to the user (from an end user’s perspective, this means the run won’t always be created immediately after the data interval ends, but only sometime later). I switched to use “is scheduled” here instead, but other suggestions are still welcomed.

> run is able to collect all the actual data within the time period. Therefore, a

Reviewer: ```suggestion
run is able to collect all the data within the time period. Therefore, a
```

run covering the data period of 2020-01-01 will not start to run until
2020-01-01 has ended, i.e. 2020-01-02 onwards.

All dates in Airflow are tied to the data interval concept in some way. The

> "logical date" (also called ``execution_date`` from previous Airflow version)

Reviewer: ```suggestion
"logical date" (also called ``execution_date`` in Airflow versions before 2.2)
```

of a DAG run, for example, usually denotes the start of the data interval, not
when the DAG is actually executed. Similarly, since the ``start_date`` argument
for the DAG and its tasks points to the same logical date, a run will only
be created after that data interval ends. So a DAG with ``@daily`` schedule and

> ``start_date`` of 2020-01-01, for example, will not be created until 2020-01-02.

Reviewer 2: What is the idea we want to deliver here? (It feels like the same information given in rows 60-63)

Author: The previous paragraph tries to establish the idea of data intervals in general, and this paragraph explains the concept underlies almost all of Airflow’s date-related configurations, including (the most commonly misunderstood) `start_date`. The intuitive interpretation from the argument name is my DAG should *start running* after this date, while in fact only the DAG’s *first data interval* starts after it. I’ve reworded both paragraphs a bit to make them read more different.

> be created after that data interval ends. So a DAG with ``@daily`` schedule and
> ``start_date`` of 2020-01-01, for example, will not be created until 2020-01-02.

Reviewer: ```suggestion
be created after that data interval ends. So a DAG with a ``@daily`` schedule and
a ``start_date`` of 2020-01-01, for example, will not be created until 2020-01-02.
```

Author: I decided to remove this example since it is too similar to the one in the previous paragraph and could be more confusing than clarifying.

Your DAG will be instantiated for each schedule along with a corresponding
DAG Run entry in the database backend.

Data Interval
-------------

Each DAG run in Airflow has an assigned "data interval" that represents the time
range it operates in. For a DAG scheduled with ``@daily``, for example, each of
its data interval would start at midnight of each day and end at midnight of the
next day.

A DAG run is scheduled *after* its associated data interval has ended, to ensure
the run is able to collect all the data within the time period. Therefore, a run
covering the data period of 2020-01-01 will not start to run until 2020-01-01

> has ended, i.e. after 2020-01-02 00:00:00.

Reviewer 3: This doesn't _have_ to be the case, it's up to the timetable to decide when it's schedulable isn't it?

Author: Technically yes; I wrote it like this to avoid going into too much “well except” stuff this early in the documentation. It’d definitely be best if we can explain this default “DAG skips one interval” behaviour in another way.

Author: I made the wording less certain on this to acount for the custom timetable edge cases. I don’t think running a DAG before its data interval ends makes enough general sense to complicate the explaination, so this is done with some “usually” and “generally”s.

Your DAG will be instantiated for each schedule along with a corresponding
DAG Run entry in the database backend.

.. _data-interval:

Data Interval
-------------

Each DAG run in Airflow has an assigned "data interval" that represents the time
range it operates in. For a DAG scheduled with ``@daily``, for example, each of
its data interval would start at midnight of each day and end at midnight of the
next day.

A DAG run is usually scheduled *after* its associated data interval has ended,
to ensure the run is able to collect all the data within the time period. In
other words, a run covering the data period of 2020-01-01 generally does not
start to run until 2020-01-01 has ended, i.e. after 2020-01-02 00:00:00.

All dates in Airflow are tied to the data interval concept in some way. The
"logical date" (also called ``execution_date`` in Airflow versions prior to 2.2)

> of a DAG run, for example, denotes the start of the data interval, not when the

Reviewer 3: I didn't think logical date and stat date were strictly tied to each other -- couldn't a custom timetable choose to do something else here?

Author: Right now logical date is hard-wired to start of data interval in the data structure. Theoratically a custom timetable can set it to something else, but it’s not exposed as a part of the API (must subclass `DataInterval`) nor explained anywhere. I think it’s best to do this as long as `executable_date` still has UNIQUE constraint to avoid people doing silly things (intentionally or not).

Reviewer 3: Ah yeah, fair.

Longer term I do want to remove the unqiue constraint on execution_date, but we aren't there yet.

DAG is actually executed.

Similarly, since the ``start_date`` argument for the DAG and its tasks points to
the same logical date, it marks the start of *the DAG's fist data interval*, not
when tasks in the DAG will start running. In other words, a DAG run will only be
scheduled one interval after ``start_date``.

.. tip::

    If ``schedule_interval`` is not enough to express your DAG's schedule,

>     logical date, or data interval, see :doc:`Customizing imetables </howto/timetable>`.

Reviewer 4: ```suggestion
    logical date, or data interval, see :doc:`Customizing timetables </howto/timetable>`.
```
